libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/wayname/MapWayname.java

```diff
@@ -0,0 +1,129 @@
+package com.mapbox.services.android.navigation.ui.v5.wayname;
+
+import android.graphics.Bitmap;
+import android.graphics.PointF;
+
+import com.mapbox.geojson.Feature;
+import com.mapbox.mapboxsdk.style.layers.Layer;
+import com.mapbox.mapboxsdk.style.layers.Property;
+import com.mapbox.services.android.navigation.ui.v5.map.FeatureInteractor;
+import com.mapbox.services.android.navigation.ui.v5.map.LayerInteractor;
+import com.mapbox.services.android.navigation.ui.v5.map.PaddingAdjustor;
+import com.mapbox.services.android.navigation.ui.v5.map.WaynameLayoutProvider;
+
+import java.util.List;
+
+import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconImage;
+import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.visibility;
+import static com.mapbox.services.android.navigation.ui.v5.map.NavigationMapboxMap.STREETS_LAYER_ID;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationConstants.MAPBOX_WAYNAME_ICON;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationConstants.MAPBOX_WAYNAME_LAYER;
+
+public class MapWayname {
+
+  private static final String NAME_PROPERTY = "name";
+  private static final int FIRST_ROAD_FEATURE = 0;
+
+  private WaynameLayoutProvider layoutProvider;
+  private LayerInteractor layerInteractor;
+  private FeatureInteractor featureInteractor;
+  private PaddingAdjustor paddingAdjustor;
+  private boolean autoQueryIsEnabled;
+  private String wayname = "";
+
+  public MapWayname(WaynameLayoutProvider layoutProvider, LayerInteractor layerInteractor,
+                    FeatureInteractor featureInteractor, PaddingAdjustor paddingAdjustor) {
+    this.layoutProvider = layoutProvider;
+    this.layerInteractor = layerInteractor;
+    this.featureInteractor = featureInteractor;
+    this.paddingAdjustor = paddingAdjustor;
+  }
+
+  public void updateWaynameWithPoint(PointF point) {
+    if (!autoQueryIsEnabled) {
+      return;
+    }
+    List<Feature> roads = findRoadLabelFeatures(point);
+    updateLayerWithRoadLabelFeatures(roads);
+  }
+
+  public void updateWaynameLayer(String wayname) {
+    Layer waynameLayer = retrieveWaynameLayer();
```

Comment:
What about passing in the `Layer` to `updateWaynameLayer` instead of having to request it here? We have the layer in the call site so no need to request it here again. Maybe we can even remove `retrieveWaynameLayer`. In general, we should avoid unnecessary indirection. Does that make sense?